src/hotspot/cpu/x86/stubGenerator_x86_64_arraycopy.cpp

```diff
@@ -717,9 +726,78 @@ address StubGenerator::generate_disjoint_copy_avx3_masked(address* entry, const
   __ leave(); // required for proper stackwalking of RuntimeStub frame
   __ ret(0);
 
+  if (MaxVectorSize == 64) {
+    __ BIND(L_copy_large);
+    arraycopy_avx3_large(to, from, temp1, temp2, temp3, temp4, count, xmm1, xmm2, xmm3, xmm4, shift);
+    __ jmp(L_finish);
+  }
   return start;
 }
 
+void StubGenerator::arraycopy_avx3_large(Register to, Register from, Register temp1, Register temp2,
+                                         Register temp3, Register temp4, Register count,
+                                         XMMRegister xmm1, XMMRegister xmm2, XMMRegister xmm3,
+                                         XMMRegister xmm4, int shift) {
+
+  // Type(shift)           byte(0), short(1), int(2),   long(3)
+  int loop_size[]        = { 256,     128,       64,      32};
+  int threshold[]        = { 4096,    2048,     1024,    512};
+
+  Label L_main_loop_large;
+  Label L_tail_large;
+  Label L_exit_large;
+  Label L_entry_large;
+  Label L_main_pre_loop_large;
+  Label L_pre_main_post_large;
+
+  if (MaxVectorSize == 64) {
+    __ BIND(L_entry_large);
+
+    __ BIND(L_pre_main_post_large);
+    // Partial copy to make dst address 64 byte aligned.
+    __ movq(temp2, to);
+    __ andq(temp2, 63);
+    __ jcc(Assembler::equal, L_main_pre_loop_large);
+
+    __ negptr(temp2);
+    __ addq(temp2, 64);
+    if (shift) {
+      __ shrq(temp2, shift);
+    }
+    __ movq(temp3, temp2);
+    copy64_masked_avx(to, from, xmm1, k2, temp3, temp4, temp1, shift, 0);
+    __ movq(temp4, temp2);
+    __ movq(temp1, count);
+    __ subq(temp1, temp2);
+
+    __ cmpq(temp1, loop_size[shift]);
+    __ jcc(Assembler::less, L_tail_large);
+
+    __ BIND(L_main_pre_loop_large);
+    __ subq(temp1, loop_size[shift]);  // whay is this here
+
+    // Main loop with aligned copy block size of 256 bytes at 64 byte copy granularity.
+    __ align32();
+    __ BIND(L_main_loop_large);
+    copy256_avx3(to, from, temp4, xmm1, xmm2, xmm3, xmm4, false, shift, 0);
+    __ addptr(temp4, loop_size[shift]);
+    __ subq(temp1, loop_size[shift]);
+    __ jcc(Assembler::greater, L_main_loop_large);
+    // fence needed because copy256_avx 3 uses non-temporal stores
+    __ sfence();
+
+    __ addq(temp1, loop_size[shift]);
+    // Zero length check.
+    __ jcc(Assembler::lessEqual, L_exit_large);
+    __ BIND(L_tail_large);
+    // Tail handling using 64 byte [masked] vector copy operations.
+    __ cmpq(temp1, 0);
+    __ jcc(Assembler::lessEqual, L_exit_large);
+    arraycopy_avx3_special_cases_256(xmm1, k2, from, to, temp1, shift,
+                                 temp4, temp3, L_entry_large, L_exit_large);
```

Comment:
When we come here to arraycopy_avx3_special_cases_256 only up to 256 bytes need to be copied so we don't need to go back to L_entry_large.

Reply:
Thanks, done